### DISCUSSION TITLE
Handling k-safety property

### DISCUSSION THREAD

**Discussion by Nivous** [seahorn/seahorn#518] [Ideas]
Hi.
I am looking to implement k-safety property using seahorn.
For that I am trying to think what is the best way to introduce the pre \ post conditions in my program.
I am not sure at all what I should use as even from a technical point of view, I am not sure how the compiler would handle those function calls and how should I implement them.
Have you considered implementing something like that and maybe have a suggestion on how to begin?

Any help would be appreciated,
Niv

**Comment by Nivous** [comment]
Still relevant. We could really use the help.

**Comment by agurfinkel** [comment]
Sorry, your email came when most of the team is on vacation / travel. The best way to communicate pre- and post-conditions is with special function calls. In the same way we communicate assumptions using `verifier.assume` call. 

It would work best if you can break your desired solution into a series of examples. If we can see what you are doing, we might be able to help.

There was a project in the past that used self-composition to convert k-safety into safety and solve it with seahorn. I don't believe it was ever merged in and seahorn has changed substantially since that time.

**Reply by Nivous** [reply]
That is exactly what we are trying to accomplish (using self-composition to convert k-safety into safety).
We are trying to implement the algorithm in the paper "Hyperproperty Verification as CHC Satisfiability" in the special case of k-saftey.

A canonical example we try to use when we think about the problem is the one introduced in section 2:
`pre(𝑎1 < 𝑎2 ∧ 𝑏1 > 𝑏2)`
`squaresSum ( int a, int b){`
`assume (0 < a < b);`
`int c=0;`
`while (a<b) {c+=a*a; a ++;}`
`return c;`
`}`
`post (𝑐1 > 𝑐2)`

We first looked at how you transformed __Verifier_Assume_ calls into special names and we think we can do the same for pre and post.
We have several issues later on though:
1. seahorn starts by compiling the c code into llvm, and then calling seapp, but llvm won't be able to compile this code since c1,c2,a1,a2,b1,b2 are not defined...
2. The paper discusses a trace of the program. We are not sure how the trace appears in seahorn, as the only output is CHCs. Is there a "nice" way to extract it?

**Reply by Nivous** [reply]
Also, could you maybe share the old code you wrote at the time? Maybe we could get some inspiration from it \ Maybe it will not be too difficult to make the necessary changes and merge it later.

**Reply by yvizel** [reply]
You do not really need the 2 copies when you write your conditions.
To handle this, you need to write your own annotations in the code, which you can then support in the pass you implement.

For example for `post(c1 > c2);`, you should introduce a relational predicate and write it as, for example, `post(_hyper_gt(c));`
Intuitively, it is a nested annotation - you first identify `post` and then analyze the argument.

I hope this helps.

Also, for post, maybe use `_hyper_post` to make sure there is no "collision" with a user defined function.

**Comment by agurfinkel** [comment]
this is more appropriate as a discussion rather than an issue

**Comment by agurfinkel** [comment]
I was about to suggest that you should talk to @yvizel , since he was the author of the previous code. Glad to see he joined the conversation :)

**Comment by Nivous** [comment]
Hi.
I am making headway in implementing the reduction that was stated in yakir's paper.
I have 2 questions:
1. For this purpose, it seems like I don't really need a query. the z3 query function requires me to have one. Should I just use True as query?
2. I am getting unknown as a result (when using True as query) even for the simplest of functions (I will attach the one I tested at the end). Is there anyway to debug \ understand why is that? Is it a timeout issue? Any help would be appreciated.

Said function:

#include "seahorn/seahorn.h"

extern int nd(void);

int main(void) {
    int a = nd();

    __hyper_pre_gt(a);
	
	if (a > 0)
		a = a + 1;
	else
		a = a + 2;

    __hyper_post_gt(a);
	
    return 0;
}

note: __hyper_post_gt(a) means that my post condition is a1>a2

**Comment by agurfinkel** [comment]
I'm not sure I understand the question. Can you be more specific what `query` you are referring to?

If this is a query for the CHC-solver, then, of course, CHC-solver needs a property to solve and query is needed. If your goal is only to generate the CHC system and not ask any questions of it, then the query is irrelevant. Otherwise, it depends on the property that you are encoding.

The argument to the query is a nullary predicate that represent a basic block. The intended meaning is whether the basic block is reachable. You cannot use `True` since it is not a nullary predicate.

To debug interaction with z3, you can use `-o` switch to generate an smt2 file and then run this file directly with z3.

For the second question, it is hard for me to answer it without seeing the CHCs that you generate. I cannot generate them myself either because I do not have access to your modifications. 

Hope this helps. I think, most likely reason, is that there is some misunderstanding and you are creating query (True) that is not supported and getting unknown as an error answer as a result.

**Reply by Nivous** [reply]
@agurfinkel 
Hi,
thanks for the reply.
My purpose is generating the CHC system, and have the solver solve it.
Basically I am interested in finding out if there is some definitions of the relations I use that can satisfy all the clauses I create.
Can you think of any query I should use to have the solver do what I want?

Thanks,
Niv
